fedimint-atomic-broadcast/src/broadcast.rs

```diff
@@ -0,0 +1,154 @@
+    outgoing_message_sender: async_channel::Sender<(Message, Recipient)>,
+    ordered_item_sender: mpsc::Sender<OrderedItem>,
+    clean_shutdown_receiver: watch::Receiver<Option<(u64, Duration)>>,
+) -> Shutdown {
+    let (network_data_sender, network_data_receiver) = async_channel::bounded(256);
+    let (signed_block_sender, signed_block_receiver) = async_channel::bounded(16);
+
+    let relay_handle = tokio::spawn(relay_messages(
+        db.clone(),
+        incoming_message_receiver,
+        outgoing_message_sender.clone(),
+        network_data_sender,
+        signed_block_sender,
+    ));
+
+    loop {
+        if let Some(signed_block) = db::load_block(&db, index).await {
+            tracing::info!("Loaded block with index {}", index);
+            if process_signed_block(signed_block, ordered_item_sender.clone())
+                .await
+                .is_err()
+            {
+                relay_handle.abort();
+                relay_handle.await.ok();
+
+                return Shutdown::MidSession(index);
+            }
+        } else {
+            tracing::info!("Run session with index {}", index);
+            let (backup_loader, backup_saver) = db::open_session(db.clone(), index).await;
+
+            let session_result = session::run(
+                index,
+                keychain.clone(),
+                backup_loader,
+                backup_saver,
+                mempool_item_receiver.clone(),
+                network_data_receiver.clone(),
+                outgoing_message_sender.clone(),
+                ordered_item_sender.clone(),
+                signed_block_receiver.clone(),
+            );
+
+            match session_result.await {
+                Ok(signed_block) => {
+                    tracing::info!("Completed session with index {}", index);
+                    db::complete_session(&db, index, signed_block.clone()).await;
+
+                    outgoing_message_sender
+                        .send((Message::Block(signed_block), Recipient::Everyone))
+                        .await
+                        .ok();
+                }
+                Err(..) => {
+                    relay_handle.abort();
+                    relay_handle.await.ok();
+
+                    return Shutdown::MidSession(index);
+                }
+            };
+        }
+
+        let clean_shutdown = clean_shutdown_receiver.borrow().to_owned();
+        if let Some((shutdown_index, shutdown_delay)) = clean_shutdown {
+            if index == shutdown_index {
+                tracing::info!("Initiate clean shutdown after index {}", index);
+
+                // prevents the relay loop from hanging if  the channels to become full
+                network_data_receiver.close();
+                signed_block_receiver.close();
+
+                // we delay the shutdown to allow lagging nodes to complete the session as well
+                sleep(shutdown_delay).await;
```

Comment:
Fixed shutdown like this are almost never correct. In practice if the delay is long enough, it might not matter, but just saying.

Reply:
I fully agree. However, this should down method would only be needed if someone would want to upgrade a federation with a coordinated switch of the fedimint binary after a session. However, since it seems currently unclear if we even want to support this upgrade method I have opted for this simple solution mainly for testing purposes or if one were to shutdown a single guardian at the end of a session without intention to upgrade.